File: app/src/main/java/de/xikolo/controllers/channels/ChannelDetailsFragment.java
package de.xikolo.controllers.channels;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.yatatsu.autobundle.AutoBundleField;

import java.util.List;

import butterknife.BindView;
import de.xikolo.App;
import de.xikolo.R;
import de.xikolo.config.GlideApp;
import de.xikolo.controllers.base.LoadingStatePresenterFragment;
import de.xikolo.controllers.course.CourseActivityAutoBundle;
import de.xikolo.controllers.login.LoginActivityAutoBundle;
import de.xikolo.controllers.main.CourseListAdapter;
import de.xikolo.models.Channel;
import de.xikolo.models.Course;
import de.xikolo.utils.SectionList;
import de.xikolo.presenters.base.PresenterFactory;
import de.xikolo.presenters.channels.ChannelDetailsPresenter;
import de.xikolo.presenters.channels.ChannelDetailsPresenterFactory;
import de.xikolo.presenters.channels.ChannelDetailsView;
import de.xikolo.views.AutofitRecyclerView;
import de.xikolo.views.SpaceItemDecoration;

public class ChannelDetailsFragment extends LoadingStatePresenterFragment<ChannelDetailsPresenter, ChannelDetailsView> implements ChannelDetailsView {

    public static final String TAG = ChannelDetailsFragment.class.getSimpleName();

    @AutoBundleField String channelId;

    // -1 do not scroll to course
    @AutoBundleField(required = false) int scrollToCoursePosition = -1;

    @BindView(R.id.layout_header) FrameLayout layoutHeader;
    @BindView(R.id.image_channel) ImageView imageChannel;
    @BindView(R.id.text_title) TextView textTitle;
    @BindView(R.id.course_list) AutofitRecyclerView courseList;

    private ChannelCourseListAdapter courseListAdapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        courseListAdapter = new ChannelCourseListAdapter(this, new CourseListAdapter.OnCourseButtonClickListener() {
            @Override
            public void onEnrollButtonClicked(String courseId) {
                presenter.onEnrollButtonClicked(courseId);
            }

            @Override
            public void onContinueButtonClicked(String courseId) {
                presenter.onCourseEnterButtonClicked(courseId);
            }

            @Override
            public void onDetailButtonClicked(String courseId) {
                presenter.onCourseDetailButtonClicked(courseId);
            }
        });

        courseList.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                return courseListAdapter.isHeader(position) ? courseList.getSpanCount() : 1;
            }
        });

        courseList.setAdapter(courseListAdapter);

        courseList.addItemDecoration(new SpaceItemDecoration(
            App.getInstance().getResources().getDimensionPixelSize(R.dimen.card_horizontal_margin),
            App.getInstance().getResources().getDimensionPixelSize(R.dimen.card_vertical_margin),
            false,
            new SpaceItemDecoration.RecyclerViewInfo() {
                @Override
                public boolean isHeader(int position) {
                    return courseListAdapter.isHeader(position);
                }

                @Override
                public int getSpanCount() {
                    return courseList.getSpanCount();
                }

                @Override
                public int getItemCount() {
                    return courseListAdapter.getItemCount();
                }
            }));
    }

    @Override
    public int getLayoutResource() {
        return R.layout.content_channel_details;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.refresh, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int videoId = item.getItemId();
        switch (videoId) {
            case R.id.action_refresh:
                presenter.onRefresh();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void setupView(Channel channel) {
        if (getActivity() instanceof ChannelDetailsActivity) {
            layoutHeader.setVisibility(View.GONE);
        } else {
            GlideApp.with(this).load(channel.imageUrl).into(imageChannel);
            textTitle.setText(channel.title);
        }

        if (courseListAdapter != null) {
            courseListAdapter.setButtonColor(channel.getColorOrDefault());
        }
    }

    @Override
    public void showCourseList(SectionList<String, List<Course>> courses) {
        if (courseListAdapter != null) {
            courseListAdapter.update(courses);
        }

        if (scrollToCoursePosition >= 0) {
            try {
                ChannelDetailsActivity activity = ((ChannelDetailsActivity) getActivity());
                if (activity != null) {
                    activity.appBarLayout.setExpanded(false);
                }

                int headerCount = 0;
                for (int i = 0; i < scrollToCoursePosition; i++) {
                    if (courseListAdapter.getItemViewType(i) == ChannelCourseListAdapter.ITEM_VIEW_TYPE_HEADER
                        || courseListAdapter.getItemViewType(i) == ChannelCourseListAdapter.ITEM_VIEW_TYPE_META) {
                        headerCount++;
                    }
                }

                courseList.smoothScrollToPosition(scrollToCoursePosition + headerCount);
                scrollToCoursePosition = -1;
            } catch (Exception ignored) {
            }
        }
    }

    @Override
    public void enterCourse(String courseId) {
        Intent intent = CourseActivityAutoBundle.builder().courseId(courseId).build(App.getInstance());
        startActivity(intent);
    }

    @Override
    public void enterCourseDetails(String courseId) {
        Intent intent = CourseActivityAutoBundle.builder().courseId(courseId).build(App.getInstance());
        startActivity(intent);
    }

    @Override
    public void openLogin() {
        Intent intent = LoginActivityAutoBundle.builder().build(App.getInstance());
        startActivity(intent);
    }

    @NonNull
    @Override
    protected PresenterFactory<ChannelDetailsPresenter> getPresenterFactory() {
        return new ChannelDetailsPresenterFactory(channelId);
    }

}

File: app/src/main/java/de/xikolo/controllers/channels/ChannelDetailsFragment.kt
package de.xikolo.controllers.channels

import android.os.Bundle
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.GridLayoutManager
import butterknife.BindView
import com.yatatsu.autobundle.AutoBundleField
import de.xikolo.App
import de.xikolo.R
import de.xikolo.config.GlideApp
import de.xikolo.controllers.base.BaseCourseListAdapter
import de.xikolo.controllers.base.ViewModelFragment
import de.xikolo.controllers.course.CourseActivityAutoBundle
import de.xikolo.controllers.helper.LoginHelper
import de.xikolo.controllers.login.LoginActivityAutoBundle
import de.xikolo.extensions.observe
import de.xikolo.extensions.observeOnce
import de.xikolo.managers.UserManager
import de.xikolo.models.Channel
import de.xikolo.models.Course
import de.xikolo.models.VideoStream
import de.xikolo.models.dao.CourseDao
import de.xikolo.network.jobs.base.NetworkCode
import de.xikolo.network.jobs.base.NetworkStateLiveData
import de.xikolo.utils.MetaSectionList
import de.xikolo.utils.extensions.openUrl
import de.xikolo.viewmodels.channel.ChannelViewModel
import de.xikolo.views.AutofitRecyclerView
import de.xikolo.views.SpaceItemDecoration

class ChannelDetailsFragment : ViewModelFragment<ChannelViewModel>() {

    companion object {
        val TAG: String = ChannelDetailsFragment::class.java.simpleName
    }

    @AutoBundleField
    lateinit var channelId: String

    // -1 do not scroll to course
    @AutoBundleField(required = false)
    var scrollToCoursePosition = -1

    @BindView(R.id.layout_header)
    internal lateinit var layoutHeader: FrameLayout

    @BindView(R.id.image_channel)
    internal lateinit var imageChannel: ImageView

    @BindView(R.id.text_title)
    internal lateinit var textTitle: TextView

    @BindView(R.id.course_list)
    internal lateinit var recyclerView: AutofitRecyclerView

    private lateinit var contentListAdapter: ChannelCourseListAdapter

    override val layoutResource = R.layout.fragment_channel_details

    override fun createViewModel(): ChannelViewModel {
        return ChannelViewModel(channelId)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        contentListAdapter = ChannelCourseListAdapter(this, object : BaseCourseListAdapter.OnCourseButtonClickListener {
            override fun onEnrollButtonClicked(courseId: String) {
                enroll(courseId)
            }

            override fun onContinueButtonClicked(courseId: String) {
                enterCourse(courseId)
            }

            override fun onDetailButtonClicked(courseId: String) {
                enterCourseDetails(courseId)
            }

            override fun onExternalButtonClicked(course: Course) {
                enterExternalCourse(course)
            }
        })

        recyclerView.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return if (contentListAdapter.isHeader(position)) recyclerView.spanCount else 1
            }
        }

        recyclerView.adapter = contentListAdapter

        recyclerView.addItemDecoration(SpaceItemDecoration(
            App.instance.resources.getDimensionPixelSize(R.dimen.card_horizontal_margin),
            App.instance.resources.getDimensionPixelSize(R.dimen.card_vertical_margin),
            false,
            object : SpaceItemDecoration.RecyclerViewInfo {
                override fun isHeader(position: Int): Boolean {
                    return contentListAdapter.isHeader(position)
                }

                override val spanCount: Int
                    get() = recyclerView.spanCount

                override val itemCount: Int
                    get() = contentListAdapter.itemCount
            }
        ))

        viewModel.channel
            .observe(viewLifecycleOwner) {
                updateView(it)
                updateContentList(
                    viewModel.buildContentList(it)
                )
            }

        viewModel.courses
            .observe(viewLifecycleOwner) {
                viewModel.channel.value?.let {
                    updateContentListAndScroll(
                        viewModel.buildContentList(it)
                    )
                }
            }
    }

    private fun updateView(channel: Channel) {
        if (activity is ChannelDetailsActivity) {
            layoutHeader.visibility = View.GONE
        } else {
            GlideApp.with(this).load(channel.imageUrl).into(imageChannel)
            textTitle.text = channel.title
            layoutHeader.visibility = View.VISIBLE
        }

        contentListAdapter.setThemeColor(channel.colorOrDefault)
        showContent()
    }

    private fun updateContentList(contents: MetaSectionList<String, Triple<String?, VideoStream?, String?>, List<Course>>) {
        contentListAdapter.update(contents)
    }

    private fun updateContentListAndScroll(contents: MetaSectionList<String, Triple<String?, VideoStream?, String?>, List<Course>>) {
        updateContentList(contents)

        if (scrollToCoursePosition >= 0) {
            try {
                val activity = activity as ChannelDetailsActivity?
                activity?.appBarLayout?.setExpanded(false)

                var headerCount = 0
                for (i in 0 until scrollToCoursePosition) {
                    if (contentListAdapter.isHeader(i)
                        || contentListAdapter.isMetaItem(i)) {
                        headerCount++
                    }
                }

                recyclerView.smoothScrollToPosition(scrollToCoursePosition + headerCount)
                scrollToCoursePosition = -1

                showContent()
            } catch (ignored: Exception) {
                showErrorMessage()
            }
        }
        showContent()
    }

    private fun enroll(courseId: String) {
        showBlockingProgress()

        val enrollmentCreationNetworkState = NetworkStateLiveData()
        enrollmentCreationNetworkState
            .observeOnce(viewLifecycleOwner) {
                when (it.code) {
                    NetworkCode.SUCCESS    -> {
                        hideAnyProgress()
                        val course = CourseDao.Unmanaged.find(courseId)
                        if (course?.accessible == true) {
                            enterCourse(courseId)
                        }
                        true
                    }
                    NetworkCode.NO_NETWORK -> {
                        hideAnyProgress()
                        showNetworkRequired()
                        true
                    }
                    NetworkCode.NO_AUTH    -> {
                        hideAnyProgress()
                        showLoginRequired()
                        openLogin()
                        true
                    }
                    else                   -> false
                }
            }

        viewModel.enroll(courseId, enrollmentCreationNetworkState)
    }

    private fun enterCourse(courseId: String) {
        if (!UserManager.isAuthorized) {
            showLoginRequired()
            openLogin()
        } else {
            val intent = CourseActivityAutoBundle.builder().courseId(courseId).build(App.instance)
            startActivity(intent)
        }
    }

    private fun enterCourseDetails(courseId: String) {
        val intent = CourseActivityAutoBundle.builder().courseId(courseId).build(App.instance)
        startActivity(intent)
    }

    private fun enterExternalCourse(course: Course) {
        activity?.openUrl(course.externalUrl)
    }

    private fun openLogin() {
        val intent = LoginHelper.loginIntent(App.instance)
        startActivity(intent)
    }

}
